### DISCUSSION TITLE
Reconsider usages of `AsyncRead` and `AsyncWrite` in puffin

### DISCUSSION THREAD

**Discussion by evenyag** [GreptimeTeam/greptimedb#4054] [General]
As mentioned in https://github.com/GreptimeTeam/greptimedb/pull/4037#issuecomment-2130755538

> We still have more work to do moving forward. For example, we need to reconsider the use of AsyncRead + AsyncSeek in our codebase, as they are not suitable for object storage services and upcoming completion-based IO.

`AsyncRead` and `AsyncWrite` are not suitable for object stores. We might need to revisit the implementation of our [puffin](https://github.com/GreptimeTeam/greptimedb/tree/main/src/puffin) crate.

This is not an urgent problem that we must solve now.

**Comment by killme2008** [comment]
@zhongzc I think we should improve it before 1.0.
